src/defs.hpp

```diff
@@ -56,21 +60,211 @@ struct LogicalLocation { // aggregate and POD type
+  const auto &morton() const { return morton_; }
 
   // operators useful for sorting
   bool operator==(LogicalLocation &ll) {
-    return ((ll.level == level) && (ll.lx1 == lx1) && (ll.lx2 == lx2) && (ll.lx3 == lx3));
+    return ((ll.level() == level_) && (ll.lx1() == lx1_) && (ll.lx2() == lx2_) &&
+            (ll.lx3() == lx3_));
   }
   static bool Lesser(const LogicalLocation &left, const LogicalLocation &right) {
-    return left.level < right.level;
+    return left.level() < right.level();
   }
   static bool Greater(const LogicalLocation &left, const LogicalLocation &right) {
-    return left.level > right.level;
+    return left.level() > right.level();
+  }
+
+  bool IsContainedIn(const LogicalLocation &container) const {
+    if (container.level() > level_) return false;
+    const std::int64_t shifted_lx1 = lx1_ >> (level_ - container.level());
+    const std::int64_t shifted_lx2 = lx2_ >> (level_ - container.level());
+    const std::int64_t shifted_lx3 = lx3_ >> (level_ - container.level());
+    return (shifted_lx1 == container.lx1()) && (shifted_lx2 == container.lx2()) &&
+           (shifted_lx3 == container.lx3());
+  }
+
+  bool Contains(const LogicalLocation &containee) const {
+    if (containee.level() < level_) return false;
+    const std::int64_t shifted_lx1 = containee.lx1() >> (containee.level() - level_);
+    const std::int64_t shifted_lx2 = containee.lx2() >> (containee.level() - level_);
+    const std::int64_t shifted_lx3 = containee.lx3() >> (containee.level() - level_);
+    return (shifted_lx1 == lx1_) && (shifted_lx2 == lx2_) && (shifted_lx3 == lx3_);
+  }
+
+  // Being a neighbor implies that you share a face, edge, or node and don't share a
+  // volume
+  bool IsNeighbor(const LogicalLocation &in) const {
+    if (in.level() < level()) return in.IsNeighbor(*this);
+    if (Contains(in)) return false; // You share a volume
+    // Only need to consider case where other block is equally or more refined than you
+    auto offset = 1 << (in.level() - level());
+    const auto shifted_lx1 = lx1_ << (in.level() - level());
+    const auto shifted_lx2 = lx2_ << (in.level() - level());
+    const auto shifted_lx3 = lx3_ << (in.level() - level());
+    const bool bx1 =
+        (in.lx1() >= (shifted_lx1 - 1)) && (in.lx1() <= (shifted_lx1 + offset));
+    const bool bx2 =
+        (in.lx2() >= (shifted_lx2 - 1)) && (in.lx2() <= (shifted_lx2 + offset));
+    const bool bx3 =
+        (in.lx3() >= (shifted_lx3 - 1)) && (in.lx3() <= (shifted_lx3 + offset));
+    return bx1 && bx2 && bx3;
+  }
+
+  LogicalLocation GetSameLevelNeighbor(int ox1, int ox2, int ox3) const {
+    return LogicalLocation(level_, lx1_ + ox1, lx2_ + ox2, lx3_ + ox3);
+  }
+
+  LogicalLocation GetParent() const {
+    if (level_ == 0) return *this;
+    return LogicalLocation(level_ - 1, lx1_ >> 1, lx2_ >> 1, lx3_ >> 1);
+  }
+
+  std::vector<LogicalLocation> GetDaughters() const {
+    std::vector<LogicalLocation> daughters;
```

Comment:
I'm sure that returning a `std::vector` here is fine. But the performance curmudgeon in me at least wants to minimize the number of resizes the vector might do internally.
```suggestion
    std::vector<LogicalLocation> daughters;
    daughters.reserve(8);
```

Reply:
I agree about possible performance issues, but I am trying not to pre-optimize any of this code. I would rather get the logic right, keep things clean, and then look for hot spots, given that the logic can be a little complex.

Reply:
I think it's probably going to be fine performance-wise. It's not like this is performance-critical code. I just suggest the `reserve` call because it's one-liner.

Reply:
added